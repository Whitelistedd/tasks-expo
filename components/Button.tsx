import {
  StyleSheet,
  TouchableOpacity,
  TouchableOpacityProps,
} from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";

export type ButtonProps = TouchableOpacityProps & {
  lightColor?: string;
  darkColor?: string;
};

export function Button({
  style,
  lightColor,
  darkColor,
  children,
  ...otherProps
}: ButtonProps) {
  const buttonColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "button"
  );

  return (
    <TouchableOpacity
      style={[{ backgroundColor: buttonColor }, styles.button, style]}
      {...otherProps}
    >
      {children}
    </TouchableOpacity>
  );
}

export const styles = StyleSheet.create({
  button: {
    paddingHorizontal: 15,
    paddingVertical: 10,
    borderRadius: 16,
  },
});
