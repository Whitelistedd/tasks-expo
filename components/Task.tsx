import { StyleSheet, ViewProps, View, TouchableOpacity } from "react-native";

import { TaskType } from "@/types/task";
import { ThemedText } from "./ThemedText";
import { CloseSVG } from "@/assets/svg/Close";
import { Colors } from "@/constants/Colors";

export type TaskProps = ViewProps & {
  lightColor?: string;
  darkColor?: string;
  task: TaskType;
  onTaskDelete: (id: number) => void;
};

export function Task({
  style,
  lightColor,
  darkColor,
  task,
  onTaskDelete,
  ...rest
}: TaskProps) {
  return (
    <View style={styles.shadow}>
      <View style={styles.container} {...rest}>
        <View style={styles.text}>
          <ThemedText testID="task_title" type="subtitle">
            {task.name}
          </ThemedText>
          {task.desc && (
            <>
              <View testID="task_line" style={styles.line} />
              <ThemedText type="desc">{task.desc}</ThemedText>
            </>
          )}
        </View>
        <TouchableOpacity
          testID="delete_button"
          onPress={() => onTaskDelete(task.id)}
          style={styles.deleteButton}
        >
          <CloseSVG />
        </TouchableOpacity>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  shadow: {
    shadowColor: "#00000050",
    shadowOffset: {
      width: 0,
      height: 2,
    },
    shadowOpacity: 0.25,
    shadowRadius: 3.84,
    elevation: 5,
    backgroundColor: "none",
    borderRadius: 16,
  },
  container: {
    borderRadius: 16,
    padding: 15,
    margin: 2,
    overflow: "hidden",
    backgroundColor: "white",
    flexDirection: "row",
    alignItems: "center",
    gap: 30,
  },
  text: {
    flex: 1,
    gap: 2,
  },
  line: {
    flex: 1,
    height: 1,
    backgroundColor: "black",
  },
  deleteButton: {
    backgroundColor: Colors.general.error,
    borderRadius: 50,
    padding: 3,
  },
});
