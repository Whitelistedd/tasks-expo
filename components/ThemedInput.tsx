import {
  StyleSheet,
  TextInput,
  TextInputProps,
  View,
  type ViewProps,
} from "react-native";

import { useThemeColor } from "@/hooks/useThemeColor";
import { ThemedText } from "./ThemedText";
import React, { ForwardedRef } from "react";

export type ThemedInputProps = TextInputProps & {
  lightColor?: string;
  darkColor?: string;
  label?: string;
};

export const ThemedInput = React.forwardRef(function ThemedInput(
  { style, lightColor, darkColor, label, ...otherProps }: ThemedInputProps,
  ref: ForwardedRef<TextInput>
) {
  const inputColor = useThemeColor(
    { light: lightColor, dark: darkColor },
    "input"
  );

  return (
    <View style={styles.container}>
      {label && <ThemedText type="desc">{label}</ThemedText>}
      <TextInput
        ref={ref}
        style={[{ borderColor: inputColor }, styles.input, style]}
        {...otherProps}
      />
    </View>
  );
});

const styles = StyleSheet.create({
  container: {
    gap: 5,
  },
  input: {
    borderWidth: 1,
    flex: 1,
    borderRadius: 8,
    padding: 8,
  },
});
