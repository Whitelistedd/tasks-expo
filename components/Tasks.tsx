import { StyleSheet, FlatList, FlatListProps } from "react-native";

import { TaskType } from "@/types/task";
import { useAppDispatch, useAppSelector } from "@/redux/store";
import { removeTask } from "@/redux/slices/tasks";
import { Task } from "./Task";

export type TasksProps = Partial<FlatListProps<TaskType>> & {
  lightColor?: string;
  darkColor?: string;
};

export function Tasks({ style, ...rest }: TasksProps) {
  const { tasks } = useAppSelector((state) => state.tasks);
  const dispatch = useAppDispatch();

  const handleTaskDelete = (id: number) => {
    dispatch(removeTask({ id }));
  };

  return (
    <FlatList
      contentContainerStyle={[styles.tasks, style]}
      data={tasks}
      renderItem={({ item }) => (
        <Task onTaskDelete={handleTaskDelete} task={item} />
      )}
      {...rest}
    />
  );
}

const styles = StyleSheet.create({
  tasks: {
    gap: 15,
    padding: 20,
  },
});
