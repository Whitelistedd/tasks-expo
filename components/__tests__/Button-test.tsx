import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import { Button } from "@/components/Button"; // Adjust the import based on your file structure
import { ThemedText } from "@/components/ThemedText";

describe("Button Component", () => {
  it("renders correctly", () => {
    const tree = render(<Button />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("given text, should render with the correct text", () => {
    const { getByText } = render(
      <Button>
        <ThemedText>Click Me</ThemedText>
      </Button>
    );

    const button = getByText("Click Me");

    expect(getByText("Click Me")).toBeTruthy();
  });

  it("given on press function, should call the onPress handler when clicked", () => {
    const handlePress = jest.fn();
    const { getByText } = render(
      <Button onPress={handlePress}>
        <ThemedText>Click Me</ThemedText>
      </Button>
    );

    const button = getByText("Click Me");

    fireEvent.press(button);
    expect(handlePress).toHaveBeenCalledTimes(1);
  });

  it("given disabled state, disabled prop should be true", () => {
    const handlePress = jest.fn();

    const { getByText } = render(
      <Button onPress={handlePress} disabled={true}>
        <ThemedText>Click Me</ThemedText>
      </Button>
    );

    const button = getByText("Click Me");

    expect(button).toBeTruthy();

    fireEvent.press(button);
    expect(handlePress).toHaveBeenCalledTimes(0);
  });
});
