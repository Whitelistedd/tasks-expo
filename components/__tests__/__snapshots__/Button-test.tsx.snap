// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component renders correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "backgroundColor": "#24A19C",
      "borderRadius": 16,
      "opacity": 1,
      "paddingHorizontal": 15,
      "paddingVertical": 10,
    }
  }
/>
`;
